contracts/MultiSigAdmin.sol

```diff
@@ -0,0 +1,667 @@
+            "MultiSigAdmin: proposal is not open"
+        );
+        _;
+    }
+
+    /**
+     * @notice Ensure that the caller can approve a given proposal
+     * @param proposalId    Proposal ID
+     */
+    modifier onlyApproverForProposal(uint256 proposalId) {
+        Proposal storage proposal = _proposals[proposalId];
+        require(
+            _types[proposal.targetContract][proposal.selector]
+                .config
+                .approvers
+                .contains(msg.sender),
+            "MultiSigAdmin: caller is not an approver"
+        );
+        _;
+    }
+
+    constructor() public Administrable() {}
+
+    /**
+     * @notice Configure who can approve and how many approvals are required for
+     * a given type of contract call
+     * @dev minApprovals must be greater than zero.
+     * @param targetContract    Address of the contract
+     * @param selector          Selector of the function in the contract
+     * @param minApprovals      Minimum number of approvals required
+     * @param approvers         List of approvers' addresses
+     */
+    function configure(
+        address targetContract,
+        bytes4 selector,
+        uint256 minApprovals,
+        address[] calldata approvers
+    ) external onlyAdmin {
+        require(
+            targetContract != address(0),
+            "MultiSigAdmin: targetContract is the zero address"
+        );
+        require(minApprovals > 0, "MultiSigAdmin: minApprovals is zero");
+        require(
+            approvers.length >= minApprovals,
+            "MultiSigAdmin: approvers fewer than minApprovals"
+        );
+
+        Configuration storage config = _types[targetContract][selector].config;
+
+        // Set minimum number of approvals
+        config.minApprovals = minApprovals;
+
+        // Set approvers
+        config.approvers.clear();
+        for (uint256 i = 0; i < approvers.length; i++) {
+            config.approvers.add(approvers[i]);
+        }
+
+        emit ConfigurationChanged(targetContract, selector, msg.sender);
+    }
+
+    /**
+     * @notice Remove configuration for a given type of contract call
+     * @param targetContract    Address of the contract
+     * @param selector          Selector of the function in the contract
+     */
+    function removeConfiguration(address targetContract, bytes4 selector)
+        external
+        onlyAdmin
+        configurationExists(targetContract, selector)
+    {
+        ContractCallType storage callType = _types[targetContract][selector];
+        Configuration storage config = callType.config;
+
+        // Reset minApprovals and approvers
+        config.minApprovals = 0;
+        config.approvers.clear();
+
+        // Close open proposals
+        // Keep the open proposal count, because it changes as the list shrinks
+        uint256 openProposalCount = callType.openProposals.length();
+        for (uint256 i = 0; i < openProposalCount; i++) {
```

Comment:
Maybe the `Configuration` struct could have a `minProposalId`, and we could set it to the `_nextProposalId` here? `_execute` could check the the `proposalId` against the `minProposalId` - so all the existing proposals would be effectively closed without looping through all the open ones.

Reply:
I am going to have to measure it, but I think removing the elements from the set will result in gas refund. But the refund happens after the call, so yes, in theory, what you said is possible, it will be expensive though.

Let me measure whether the gas refund is preferable over just invalidating them (with the minProposal idea you have) and forgoing the refund.

Reply:
Not counting the control flow and memory operations, updating the minProposal will cost 5,000 (1 storage update), removing each item in the set causes 2 updates (5,000 + 5,000) and 2 removals (-15,000 + -15,000) so we will get back 20,000 gas. I will measure it to be sure, but it seems like enforcing max open proposals limit would be the most gas efficient?

Reply:
I did some tests and it does seem like setting the minProposal is still much cheaper. I also forgot about emitting the `ProposalClosed` event.

There are two options:
1. Go without `ProposalClosed` event and set `minProposalId` to invalid proposals.
2. Make it possible to set an upper limit on the number of open proposals for each configuration.

Neither option is perfect though.
* `ProposalClosed` event is useful and it's nice to be able to get a full audit log just by printing all log events this contract has emitted.
* Setting a limit for the quantity open proposals may be cumbersome in some use cases, but it's per configuration and needing to queue up a ton of proposals for a specific type of contract call would be rare.

Reply:
After chatting with @eztierney, I've decided to go with the following:

* Introduce a `ProposalsInvalidated(targetContract, selector, by)` event.
* Use `minProposalId` to mass-invalidate proposals.

This is the cheapest approach in terms of gas, and we don't have to restrict the quantity of open proposals. The downside is that the new event won't give you the IDs of proposals that are invalidated, but we can still figure out it from the stream of events with a little more work anyway.

Reply:
I realized as I was implementing the above that because this approach could potentially let the `openProposals` set keep growing with lots of invalidated proposals that are only filtered out at runtime, it will, at some point, make `getOpenProposals` fail. View functions don't consume any gas when called outside of a transaction, but AFAIK still fail if gas count > block gas limit. I think I will just go with the max quantity approach.

Reply:
Technically, a lot of these view functions this contract provides like `getOpenProposals` and `getApprovals` and the use of `EnumerableSet` are not strictly needed, as it is possible to infer most of the info just by processing the stream of log events, but I was trying to prioritize the easy of use over optimizing gas cost. With a GUI-frontend that does the processing of log events, we might able to do away with a lot of these niceties though. Let me revisit that in the future.